storage/index.go

```diff
@@ -21,61 +21,260 @@
+		// NB(prateek): skip calling any of the `OnIndexFinalize` methods
+		// on the provided inserts to terminate quicker during shutdown.
+		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return errDbIndexUnableToWriteClosed
+	}
+
+	var numErr int64
+	for _, insert := range inserts {
+		// FOLLOWUP(prateek): need to query before insert to ensure no duplicates && add test
+		err := i.active.segment.Insert(insert.doc)
+		if err != nil {
+			numErr++
+		} else {
+			insert.fns.OnIndexSuccess(i.active.expiryTime)
+		}
+		// NB: we need to release held resources so we un-conditionally execute the Finalize.
+		insert.fns.OnIndexFinalize()
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
+	}
+
+	return nil
 }
 
 func (i *dbIndex) Write(
-	ctx context.Context,
 	namespace ident.ID,
 	id ident.ID,
-	tags ident.TagIterator,
+	tags ident.Tags,
+	fns indexInsertLifecycleHooks,
 ) error {
-	return errIndexingNotImplemented
+	d, err := i.doc(namespace, id, tags)
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		i.metrics.InsertAfterClose.Inc(1)
+		fns.OnIndexFinalize()
+		return errDbIndexUnableToWriteClosed
+	}
+
+	async := i.insertMode == dbIndexInsertAsync
+	wg, err := i.insertQueue.Insert(d, fns)
+	i.RUnlock()
+
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	// once the write has been queued in the indexInsertQueue, it assumes
+	// responsibility for calling the lifecycle resource hooks.
+
+	if !async {
+		wg.Wait()
+	}
+
+	return nil
 }
 
 func (i *dbIndex) Query(
 	ctx context.Context,
 	query index.Query,
 	opts index.QueryOptions,
 ) (index.QueryResults, error) {
-	return index.QueryResults{}, errIndexingNotImplemented
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		return index.QueryResults{}, errDbIndexUnableToQueryClosed
+	}
+
+	// FOLLOWUP(prateek): push down QueryOptions to restrict results
+	iter, err := i.active.segment.Query(query.Query)
+	i.RUnlock()
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	return index.QueryResults{
+		Iterator:   index.NewIterator(iter, i.opts),
+		Exhaustive: true,
+	}, nil
 }
 
-type databaseIndexWriter interface {
-	Write(
-		ctx context.Context,
-		namespace ident.ID,
-		id ident.ID,
-		tags ident.TagIterator,
-	) error
+func (i *dbIndex) isOpenWithRLock() bool {
+	return i.state == dbIndexStateOpen
+
 }
 
-type dbIndexNoOp struct{}
+func (i *dbIndex) Close() error {
+	i.Lock()
+	defer i.Unlock()
+	state := i.state
+	if state != dbIndexStateOpen {
+		return errDbIndexAlreadyClosed
+	}
 
-func (n dbIndexNoOp) Write(context.Context, ident.ID, ident.ID, ident.TagIterator) error {
-	return nil
+	i.state = dbIndexStateClosed
+	return i.insertQueue.Stop()
+}
+
+func (i *dbIndex) doc(ns, id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameNamespace,
```

Comment:
Even if this namespace is really long, the index will end up compressing the re-used namespace pretty well right?

Reply:
Yeah either implementation of segment will keep a bitset of documents per unique (fieldname, fieldvalue)

Reply:
Not sure I'm totally following. I guess my question was if you insert a million document, we won't have a million different ident.ID's referring to the same namespace right?